File: app/src/main/java/tech/hamid/bookery/LoginActivity.java
package tech.hamid.bookery;
import androidx.annotation.NonNull;
import androidx.appcompat.app.ActionBar;
import androidx.appcompat.app.AlertDialog;
import androidx.appcompat.app.AppCompatActivity;

import android.content.DialogInterface;
import android.content.Intent;
import android.os.Bundle;
import android.text.TextUtils;
import android.view.MenuItem;
import android.view.View;
import android.widget.Button;
import android.widget.EditText;
import android.widget.ProgressBar;
import android.widget.TextView;
import android.widget.Toast;

import com.google.android.gms.tasks.OnCompleteListener;
import com.google.android.gms.tasks.OnFailureListener;
import com.google.android.gms.tasks.OnSuccessListener;
import com.google.android.gms.tasks.Task;
import com.google.firebase.auth.AuthResult;
import com.google.firebase.auth.FirebaseAuth;

public class LoginActivity extends AppCompatActivity {

    EditText memail, mpassword;
    Button mloginbtn;
    TextView mcreatebtn, forgotPasswordLink;
    ProgressBar progressBar;
    FirebaseAuth fAuth;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_login);

        memail = findViewById(R.id.email);
        mpassword = findViewById(R.id.password);
        progressBar = findViewById(R.id.progressBar2);
        fAuth = FirebaseAuth.getInstance();
        mloginbtn = findViewById(R.id.Loginbtn);
        mcreatebtn = findViewById(R.id.createtext);
        forgotPasswordLink = findViewById(R.id.forgotpassword);

        ActionBar actionBar = getSupportActionBar();
        if (actionBar != null) {
            actionBar.setDisplayHomeAsUpEnabled(true);
        }

        mloginbtn.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                String email = memail.getText().toString().trim();
                String password = mpassword.getText().toString().trim();

                if(TextUtils.isEmpty(email)) {
                    memail.setError("Email is required.");
                    return;
                }

                if(TextUtils.isEmpty(password)) {
                    mpassword.setError("password is required.");
                    return;
                }

                if(password.length() < 6) {
                    mpassword.setError("password must be greater or equal to six characters.");
                    return;
                }
                progressBar.setVisibility(View.VISIBLE);

                fAuth.signInWithEmailAndPassword(email, password).addOnCompleteListener(new OnCompleteListener<AuthResult>() {
                    @Override
                    public void onComplete(@NonNull Task<AuthResult> task) {
                        progressBar.setVisibility(View.GONE);
                        if(task.isSuccessful()) {
                            Toast.makeText(LoginActivity.this, "Logged in successfully", Toast.LENGTH_SHORT).show();
                            Intent intent = new Intent(getApplicationContext(), MainActivity.class);
                            startActivity(intent);
                            finishAffinity();
                        }
                        else {
                            Toast.makeText(LoginActivity.this,"Error! " + task.getException().getMessage(), Toast.LENGTH_SHORT).show();
                        }
                    }
                });
            }
        });

        mcreatebtn.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                startActivity(new Intent(getApplicationContext(), RegisterActivity.class));
            }
        });


        forgotPasswordLink.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                final EditText resetMail = new EditText(v.getContext());
                final AlertDialog.Builder passwordResetDialog = new AlertDialog.Builder(v.getContext());
                passwordResetDialog.setTitle("Reset Password ?");
                passwordResetDialog.setMessage("Enter your Email to recieve reset link.");
                passwordResetDialog.setView(resetMail);

                passwordResetDialog.setPositiveButton("yes", new DialogInterface.OnClickListener() {
                    @Override
                    public void onClick(DialogInterface dialog, int which) {
                        String mail = resetMail.getText().toString();
                        fAuth.sendPasswordResetEmail(mail).addOnSuccessListener(new OnSuccessListener<Void>() {
                            @Override
                            public void onSuccess(Void aVoid) {
                                Toast.makeText(LoginActivity.this, "Reset link has been sent to your Email.", Toast.LENGTH_SHORT).show();
                            }
                        }).addOnFailureListener(new OnFailureListener() {
                            @Override
                            public void onFailure(@NonNull Exception e) {
                                Toast.makeText(LoginActivity.this, "Error! Reset Link is not sent" +e.getMessage(), Toast.LENGTH_SHORT).show();
                            }
                        });
                    }
                });

                passwordResetDialog.setNegativeButton("No", new DialogInterface.OnClickListener() {
                    @Override
                    public void onClick(DialogInterface dialog, int which) {

                    }
                });
                passwordResetDialog.create().show();
            }
        });
    }

//    @Override
//    public boolean onOptionsItemSelected(@NonNull MenuItem item) {
//        Log.d("anjum", "onOptionsItemSelected");
//        Log.d("anjum", String.valueOf(item.getItemId()));
//        Log.d("anjum", String.valueOf(android.R.id.home));
//        return super.onOptionsItemSelected(item);
//    }

    @Override
    public boolean onOptionsItemSelected(@NonNull MenuItem item) {
        switch (item.getItemId()) {
            case android.R.id.home:
                finish();
                return true;
        }
        return super.onOptionsItemSelected(item);
    }

}

File: app/src/main/java/tech/hamid/bookerykotlin/LoginActivity.kt
package tech.hamid.bookerykotlin

import android.content.DialogInterface
import android.content.Intent
import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import android.text.TextUtils
import android.view.MenuItem
import android.view.View
import android.widget.*
import androidx.appcompat.app.AlertDialog
import com.google.firebase.auth.FirebaseAuth

class LoginActivity : AppCompatActivity() {

    lateinit var mEmail : EditText
    lateinit var mPassword : EditText
    lateinit var mLoginBtn : Button
    lateinit var mCreateBtn : TextView
    lateinit var forgotPasswordLink : TextView
    lateinit var progressBar: ProgressBar
    lateinit var fAuth : FirebaseAuth

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_login)

        mEmail = findViewById(R.id.email)
        mPassword = findViewById(R.id.password)
        progressBar = findViewById(R.id.progressBar2)
        fAuth = FirebaseAuth.getInstance()
        mLoginBtn = findViewById(R.id.Loginbtn)
        mCreateBtn = findViewById(R.id.createtext)
        forgotPasswordLink = findViewById(R.id.forgotpassword)

        var actionBar = supportActionBar
        if (actionBar != null) {
            actionBar.setDisplayHomeAsUpEnabled(true)
        }

        mLoginBtn.setOnClickListener(object : View.OnClickListener {
            override fun onClick(v: View?) {
                var email: String = mEmail.text.toString().trim()
                var password: String = mPassword.text.toString().trim()

                if (TextUtils.isEmpty(email)) {
                    mEmail.error = "Email is required."
                    return
                }
                if (TextUtils.isEmpty(password)) {
                    mPassword.error = "Password is required"
                    return
                }
                if (password.length < 6) {
                    mPassword.error = "password must be greater or equal to six characters."
                    return
                }

                progressBar.visibility = View.VISIBLE

                fAuth.signInWithEmailAndPassword(email, password).addOnCompleteListener { task ->
                    progressBar.visibility = View.GONE
                    if (task.isSuccessful) {
                        Toast.makeText(this@LoginActivity, "Logged in successfully", Toast.LENGTH_SHORT).show()
                        var intent = Intent(applicationContext, MainActivity::class.java)
                        startActivity(intent)
                        finishAffinity()
                    } else {
                        Toast.makeText(this@LoginActivity, "Couldn't Login, please try again", Toast.LENGTH_SHORT).show()
                    }
                }
            }
        })

        mCreateBtn.setOnClickListener {
            finish()
        //    startActivity(Intent(applicationContext, RegisterActivity::class.java))
        }

        forgotPasswordLink.setOnClickListener(object : View.OnClickListener {
            override fun onClick(v: View?) {
                val resetMail = EditText(v?.context)
                val passwordResetDialog = v?.let { AlertDialog.Builder(it.context) }

                if (passwordResetDialog != null) {
                    passwordResetDialog.setTitle("Reset Password ?")
                    passwordResetDialog.setMessage("Enter your Email to recieve reset link.")
                    passwordResetDialog.setView(resetMail)
                }

                passwordResetDialog!!.setPositiveButton("yes") { dialog, which ->
                    val mail = resetMail.text.toString()
                    fAuth.sendPasswordResetEmail(mail).addOnSuccessListener {
                        Toast.makeText(
                            this@LoginActivity,
                            "Reset link has been sent to your Email.",
                            Toast.LENGTH_SHORT
                        ).show()
                    }.addOnFailureListener { e ->
                        Toast.makeText(
                            this@LoginActivity,
                            "Error! Reset Link is not sent" + e.message,
                            Toast.LENGTH_SHORT
                        ).show()
                    }
                }
                passwordResetDialog.setNegativeButton(
                    "No",
                    DialogInterface.OnClickListener { dialog, which -> })
                passwordResetDialog.create().show()
            }
        })
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            android.R.id.home -> {
                finish()
                return true
            }
        }
        return super.onOptionsItemSelected(item)
    }
}
